mesh/mesh_readers.cpp

```diff
@@ -1291,6 +1291,64 @@ void Mesh::ReadGmshMesh(std::istream &input)
          MFEM_CONTRACT_VAR(elem_domain);
 
       } // section '$Elements'
+      else if (buff == "$Periodic") // Reading master/slave node pairs
+      {
+         curved = 1;
+         read_gf = 0;
+         spaceDim = 3;
+
+         Array<int> v2v(NumOfVertices);
+         for (int i = 0; i < v2v.Size(); i++)
+         {
+            v2v[i] = i;
+         }
+         int num_per_ent;
+         int num_nodes;
+         int slave, master;
+         input >> num_per_ent;
+         getline(input, buff); // Read end-of-line
+         for (int i = 0; i < num_per_ent; i++)
+         {
+            getline(input, buff); // Read and ignore entity dimension and tags
+            getline(input, buff); // Read and ignore affine mapping
+            // Read master/slave vertex pairs
+            input >> num_nodes;
+            for (int j=0; j<num_nodes; j++)
+            {
+               input >> slave >> master;
+               v2v[slave - 1] = master - 1;
+            }
+            getline(input, buff); // Read end-of-line
+         }
+
+         // Convert nodes to discontinuous GridFunction
+         this->SetCurvature(1, true, Dim, Ordering::byVDIM);
+
+         // Renumber elements to remove slave vertices
```

Comment:
Could you elaborate what this does? Lets say we are looking at a rectangle (0,1)x(0,1) which is periodic at x=0 and x=1.

Reply:
I am having troubles running ex1 with mesh generated by the following geo file:

SetFactory("OpenCASCADE");
  
R = 2;
Torus(2) = {3,0,0, 1.5, 1, Pi/3};

pts() = PointsOf{ Volume{2}; };

Characteristic Length{ pts() } = 0.4;

// Characteristic Length{ 2 } = 0.01;

// Find bottom and top surfaces using a bounding box search:
e = 1e-6;
bottom() = Surface In BoundingBox{-e,-e,-e, R+e,R+e,e};
top() = Surface In BoundingBox{-e,-e,R-e, R+e,R+e,R+e};

// Set a rotation periodicity constraint:
Periodic Surface{9} = {8} Rotate{{0,0,1}, {3,0,0}, Pi/3 };

I think a test example in necessary.

Reply:
@jandrej , are you suggesting I expand the comment for this line to explain MFEM's strategy for handling periodic meshes?  I can certainly do that if you like.

Reply:
@bslazarov , an example is a good idea.  I can add something simple.  I tried your geo file in Gmsh 4.5.6 and got an error regarding surfaces 9 and 8 not being found.  I don't work with Gmsh myself.  The examples I was working from were provided to me by one of the researchers in FES.

Reply:
I was just curious about the "remove slave vertices" comment.

Reply:
As I look at it again I see that comment is not clear.  I'll reword it.  What happens in the following loops is that the element and boundary element connectivity are altered to replace all references to the "slave" vertices with references to the corresponding "master" vertices.  This effectively stitches the two periodic surfaces together topologically.

Reply:
I have messed up the last line :-)

Here is the correct one:
// Set a rotation periodicity constraint:
Periodic Surface{3} = {2} Rotate{{0,0,1}, {3,0,0}, Pi/3 };

Reply:
Hi, @bslazarov , the trouble was not with the periodicity but with the element and boundary attributes which are equal to zero by default.  Adding the `Physical [Surface|Volume] ...` commands towards the end of the geo file seems to help.

```
SetFactory("OpenCASCADE");

R = 2;
Torus(1) = {3,0,0, 1.5, 1, Pi};

pts() = PointsOf{ Volume{1}; };

Characteristic Length{ pts() } = 0.4;

// Set a rotation periodicity constraint:
Periodic Surface{3} = {2} Rotate{{0,0,1}, {3,0,0}, Pi };

// Tag surfaces and volumes with positive integers
Physical Surface(1) = {1};
Physical Surface(2) = {2};
Physical Surface(3) = {3};
Physical Volume(1) = {1};

// Generate 3D mesh
Mesh 3;
```

Reply:
Actually there was a periodicity issue as well.  If the characteristic length is too large there will be fewer than three edges connecting the periodic surfaces and MFEM cannot always distinguish such edges.

Reply:
Thanks

Reply:
@mlstowell  - it seems that I am getting the same errors again, even if I tag the boundaries. I am getting the following errors:

MFEM abort: Gmsh file : vertices indices are not unique
 ... in function: void mfem::Mesh::ReadGmshMesh(std::istream&, int&, int&)
 ... in file: /home/blaz/develop/mark/mfem/mesh/mesh_readers.cpp:952

 The error appear for non-periodic mesh also. I am using 4.1 format for gmsh - let me know if that is the issue.

Reply:
@jandrej , working with @bslazarov I just added 2D and 3D example meshes to the `data` subdirectory and new sample runs to exercise the new code.